### PR TITLE
Bugfix: fix chan leak when stop error

### PR DESCRIPTION
**Pull request by cwdsuzhou**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fix chan leak when stop error

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/ok-to-test
/priority backlog

**Review by mattjmcnaughton — CHANGES_REQUESTED** [review]
Thanks as always for looking to fix leaks @cwdsuzhou 

Unfortunately, I'm not sure this change is safe to make, as it introduces the risk of a `panic` caused by attempting to close `c` twice.

For example, imagine that after 11 seconds, we close c bc of the timeout. Then, 5 seconds later, the `w.wg.Wait()` call returns and we try and close c again. There will be a panic.

I _think_ that to solve this we need some form of check that `c` is only going to be closed once. https://groups.google.com/forum/#!topic/golang-nuts/rhxMiNmRAPk has some interesting discussion on this topic :)

**Comment by cwdsuzhou** [comment]
> Thanks as always for looking to fix leaks @cwdsuzhou

Thanks @mattjmcnaughton , I am not alway looking for leaks, just because I feel there is something incorrect in the function `Stop`

> Unfortunately, I'm not sure this change is safe to make, as it introduces the risk of a `panic` caused by attempting to close `c` twice.
> 
> For example, imagine that after 11 seconds, we close c bc of the timeout. Then, 5 seconds later, the `w.wg.Wait()` call returns and we try and close c again. There will be a panic.

When this  function `Stop` exit, the go routine would also exit, so the panic may not occur.

**Comment by mattjmcnaughton** [comment]
> > Thanks as always for looking to fix leaks @cwdsuzhou
> 
> Thanks @mattjmcnaughton , I am not alway looking for leaks, just because I feel there is something incorrect in the function `Stop`
> 
> > Unfortunately, I'm not sure this change is safe to make, as it introduces the risk of a `panic` caused by attempting to close `c` twice.
> > For example, imagine that after 11 seconds, we close c bc of the timeout. Then, 5 seconds later, the `w.wg.Wait()` call returns and we try and close c again. There will be a panic.
> 
> When this function `Stop` exit, the go routine would also exit, so the panic may not occur.

I agree that in the current model, there's a low chance of `c` being closed twice. I was thinking more of ensuring we have protection against this going forward, in case more code is added to the Stop function.

**Comment by cwdsuzhou** [comment]
@mattjmcnaughton thanks for your clarification, you are right, there my be a low chance of `c` being closed twice. 

I  have updated the pr and use sync.Once to make sure the chan `c` would be closed only once.  

PTAL, thanks!

File: pkg/kubelet/util/pluginwatcher/plugin_watcher.go [hunk 1]
```diff
@@ -161,14 +161,17 @@ func (w *Watcher) Stop() error {
 	close(w.stopCh)
 
 	c := make(chan struct{})
+	var once sync.Once
+	closeFunc := func() { close(c) }
```
Comment (cwdsuzhou) [inline pkg/kubelet/util/pluginwatcher/plugin_watcher.go]: @mattjmcnaughton Add sync.Once to keep close() would be called only once

Comment (mattjmcnaughton) [reply pkg/kubelet/util/pluginwatcher/plugin_watcher.go]: Thanks :)

**Review by mattjmcnaughton — APPROVED** [review]
/lgtm

Thanks for adding the `sync.Once`.

File: pkg/kubelet/util/pluginwatcher/plugin_watcher.go (hunk 1 above)
Comment (mikedanese) [reply pkg/kubelet/util/pluginwatcher/plugin_watcher.go]: close can be called on a chan as many times as you want. The once is not necessary.

**Comment by mikedanese** [comment]
Problem with existing code: The use of waitgroup doesn't make sense here. Why is a hold being added to multiple times in the same goroutine? wg could be turned into a chan to simplify this significantly.

**Comment by cwdsuzhou** [comment]
> Problem with existing code: The use of waitgroup doesn't make sense here. Why is a hold being added to multiple times in the same goroutine? wg could be turned into a chan to simplify this significantly.
> 
> Edit: nevermind, missed the wg hold in traverse, but still claim that a chan would be simpler.

Thanks @mikedanese ,  I also think there are some optimization could be done, I am trying to optimize it.

**Comment by cwdsuzhou** [comment]
@mikedanese I  found since https://github.com/kubernetes/kubernetes/pull/75110 merged. WaitGroup could be removed and `Stop` function could be simplified.

I  have sent a PR https://github.com/kubernetes/kubernetes/pull/77907

**Pull request merged**